### DISCUSSION TITLE
Brainstorming on `Reporter` interface definition

### DISCUSSION THREAD

**Discussion by tarkatronic** [underdog-tech/vulnbot#27] [General]
Working on #5, I'm trying to define an interface for a `Reporter`. The first concrete implementation will be a `SlackReporter`. Other ones I have in mind immediately are:

* ConsoleReporter
* EmailReporter
* DatadogReporter

What I'm thinking right now is:

```go
type Reporter interface {
	SendSummaryReport(header string, numRepos int, report VulnerabilityReport, wg sync.WaitGroup)
	SendTeamReports(teamReports map[string]map[string]VulnerabilityReport, wg sync.WaitGroup)
}
```

This basically matches with how we're currently sending our messages directly to Slack. I think the ideal calling code would be something like:

```go
var wg sync.WaitGroup
for reporter := range reporters {
    reporter.SendSummaryReport("Dependabot Report for ...", len(allRepos), vulnSummary, wg)
    reporter.SendTeamReports(teamReports, wg)
}
wg.Wait()
```

How does this look? What am I missing here?

**Comment by JoseAngel1196** [comment]
The interface definition looks good. I was just thinking of one thing that might be worth considering: Async vs Sync. Currently, The methods in the Reporter interface are blocking, which means that calling code will be blocked until all reports are sent. We could consider making the reporting methods asynchronous by using channels or goroutines to send reports concurrently.

**Reply by tarkatronic** [reply]
Ah, right I totally forgot to add that. My example was supposed to look like

```go
var wg sync.WaitGroup
for reporter := range reporters {
    go reporter.SendSummaryReport("Dependabot Report for ...", len(allRepos), vulnSummary, wg)
    go reporter.SendTeamReports(teamReports, wg)
}
wg.Wait()
```
Adding the `go` is basically all we need to make this async/coroutines 😅

**Reply by tarkatronic** [reply]
The idea is that we pass that `WaitGroup` to the reporters, and inside the implementations they would all have something like

```go
func (s *SlackReporter) SendSummaryReport(header string, numRepos int, report VulnerabilityReport, wg sync.WaitGroup) {
    wg.Add(1)
    defer wg.Done()
    // Send the message
}
```
And the `wg.Wait()` at the end makes sure we wait until all of the reports are sent.

**Reply by JoseAngel1196** [reply]
Wow, it looks so easy! I'm glad to see that it's part of the roadmap for this task. 😄

**Comment by tarkatronic** [comment]
Alright, I've got all of this implemented in #39. I had to modify the interface a little bit. But I think it's going to work fairly well, for now.

**Comment by tarkatronic** [comment]
Closing this discussion now, as #39 has been merged!
